pkg/p2p/server.go

```diff
@@ -0,0 +1,872 @@
+		// no existing peer
+		m.peers[streamMeta.SenderId] = newCDCPeer(streamMeta.SenderId, streamMeta.Epoch, sender)
+		m.peerLock.Unlock()
+	} else {
+		m.peerLock.Unlock()
+		// there is an existing peer
+		if peer.Epoch > streamMeta.Epoch {
+			log.Warn("incoming connection is stale",
+				zap.String("sender-id", streamMeta.SenderId),
+				zap.String("addr", clientIP),
+				zap.Int64("epoch", streamMeta.Epoch))
+
+			// the current stream is stale
+			return cerror.ErrPeerMessageStaleConnection.GenWithStackByArgs(streamMeta.Epoch /* old */, peer.Epoch /* new */)
+		} else if peer.Epoch < streamMeta.Epoch {
+			err := cerror.ErrPeerMessageStaleConnection.GenWithStackByArgs(peer.Epoch /* old */, streamMeta.Epoch /* new */)
+			m.deregisterPeer(ctx, peer, err)
+			m.peerLock.Lock()
+			m.peers[streamMeta.SenderId] = newCDCPeer(streamMeta.SenderId, streamMeta.Epoch, sender)
+			m.peerLock.Unlock()
+		} else {
+			log.Warn("incoming connection is duplicate",
+				zap.String("sender-id", streamMeta.SenderId),
+				zap.String("addr", clientIP),
+				zap.Int64("epoch", streamMeta.Epoch))
+
+			return cerror.ErrPeerMessageDuplicateConnection.GenWithStackByArgs(streamMeta.Epoch)
+		}
+	}
+
+	return nil
+}
+
+func (m *MessageServer) scheduleTask(ctx context.Context, task interface{}) error {
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case m.taskQueue <- task:
+	default:
+		return cerror.ErrPeerMessageTaskQueueCongested.GenWithStackByArgs()
+	}
+	return nil
+}
+
+func (m *MessageServer) scheduleTaskBlocking(ctx context.Context, task interface{}) error {
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case m.taskQueue <- task:
+	}
+	return nil
+}
+
+// SendMessage implements the gRPC call SendMessage.
+func (m *MessageServer) SendMessage(stream p2p.CDCPeerToPeer_SendMessageServer) error {
+	packet, err := stream.Recv()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := m.verifyStreamMeta(packet.Meta); err != nil {
+		msg := errorToRPCResponse(err)
+		_ = stream.Send(&msg)
+		return errors.Trace(err)
+	}
+
+	metricsServerStreamCount := serverStreamCount.With(prometheus.Labels{
+		"from": packet.Meta.SenderAdvertisedAddr,
+	})
+	metricsServerStreamCount.Add(1)
+	defer metricsServerStreamCount.Sub(1)
+
+	streamHandle := m.newStreamHandle(packet.Meta)
+	ctx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+	errg, ctx := errgroup.WithContext(ctx)
+
+	errg.Go(func() error {
+		defer streamHandle.Close()
+		if err := m.receive(stream, streamHandle); err != nil {
+			log.Warn("peer-to-peer message handler error", zap.Error(err))
+			select {
+			case <-ctx.Done():
+				log.Warn("error receiving from peer", zap.Error(ctx.Err()))
+				return errors.Trace(ctx.Err())
+			case streamHandle.sendCh <- errorToRPCResponse(err):
+			default:
+				log.Warn("sendCh congested, could not send error", zap.Error(err))
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	})
+	errg.Go(func() error {
```

Comment:
Looks like this goroutine is unnecessary, we can run the for loop in `SendMessage`.

Reply:
If we did this, we would not be able to select on `m.closeCh`, which makes it impossible to terminate the stream properly if `stream.Send` is blocked by gRPC flow control.